### PR TITLE
IO implementation

### PR DESCRIPTION
**Pull request by xplosunn**

### CODE REVIEW

File: core/src/main/scala/com/codacy/analysis/core/tools/Tool.scala [hunk 1]
```diff
@@ -79,17 +80,24 @@ class Tool(runner: ToolRunner, defaultRunTimeout: Duration)(private val plugin:
...
+            FileHelper.relativePath(sourceDirectory.appendPrefix(r.filename)),
+            Issue.Message(r.message),
+            r.level,
+            r.category,
+            LineLocation(r.line)))(collection.breakOut): Set[ToolResult]) ++
+          res.failedFiles.map(fe => FileError(FileHelper.relativePath(fe), "Failed to analyse file."))
+      }
     }
+
+    for {
```
Comment (rtfpessoa) [inline core/src/main/scala/com/codacy/analysis/core/tools/Tool.scala]: Why do you need this?

File: build.sbt [hunk 2]
```diff
@@ -47,6 +47,7 @@ lazy val codacyAnalysisCore = project
       Dependencies.fansi,
       Dependencies.scalajHttp,
       Dependencies.jGit,
+      Dependencies.zio,
       Dependencies.cats) ++
```
Comment (rtfpessoa) [inline build.sbt]: This is scary 😨

**Comment by bmbferreira** [comment]
@xplosunn At first sight, this kind of approach seems well suited and I will spend some more time navigating the code and try to get more familiar with the zio library. 

Meanwhile, I also took a look at the compilation times in CircleCi to check the impact of this change and noticed that it increased from [00:57](https://circleci.com/gh/codacy/codacy-analysis-cli/3975) to [01:41](https://circleci.com/gh/codacy/codacy-analysis-cli/3948).

**Comment by xplosunn** [comment]
Rebased and opened another PR: #134

**Pull request closed without merging**